### DISCUSSION TITLE
Prediction formula with indices

### DISCUSSION THREAD

**Discussion by tmeeha** [inlabru-org/inlabru#212] [Q&A] [answered]
Hi all,

I have a spatially varying coefficient model running well in inlabru and am wondering how fancy I can get with the prediction formula.  Right now I am running:

# model
form <- state_count ~ Intercept(1) +  
  effort_eff(log_hours, model="linear") +
  year_eff(year_idx, model="linear") + 
  state_int(state_idx, model="bym2", graph=states_weights, hyper=pc_prec) + 
  state_year_dev(state_idx, year_idx, model="besag", graph=states_weights, hyper=pc_prec) +
  state_effort_dev(state_idx, log_hours, model="besag", graph=states_weights, hyper=pc_prec)
  
# estimate model
res1 <- bru(form, data=cbc2, family = "nbinomial", 
            options=list(verbose=TRUE,
                         control.compute=list(cpo=TRUE, waic=TRUE, config=TRUE),
                         control.predictor=list(compute=TRUE, link=1)))

# make predictions
pred1 <- cbc2 %>% 
  select(year, state, state_idx, year_idx, state_hours, log_hours, state_count) %>% 
  distinct() %>% 
  group_by(state) %>% mutate(log_hours=log(mean(state_hours))) %>% ungroup() %>% 
  predict(
    object=res1, 
    formula=~exp(Intercept +  
                   effort_eff + year_eff +
                   state_int + 
                   state_effort_dev +
                   state_year_dev), 
    n.samples=5000
  )

This gives me an output table with 58 states times 10 years = 580 rows. All is well. I am wondering if I can do things like get the sum of states PER year and then get the difference between the LAST and FIRST year. In my mind this would entail a formula that allows indexing. Is this type of thing possible? Or should I move on to inla.posterior.sample() or maybe generate()?

Thanks, 
Tim

**Comment by finnlindgren** [comment, marked as answer]
Short answer: you can get very fancy with the prediction formula. The "formula" argument just treats the RHS of the `~` as a general R expression, which allows things like
```
~ {
  some_thing <- my_effect[c(2,6,42)]
  some_other_thing <- my_effect_eval(main = 1:3, group = some_variable) * other_effect_latent
  list(A = some_thing, B = some_other_thing)
  }
```
giving a list of data.frames names A and B as the output.
If the result of the expression is a regular vector, you get just a single data.frame output.
What you can do is mostly limited by your R coding imagination...

Btw, I've made various updates to https://inlabru-org.github.io/inlabru/articles/svc.html to use the newer methods for meshes and geometry in general.

The longer answer is that `predict()` calls `generate()`, that first calls `inla.posterior.sample()`, and then extracts the latent state values/vectors and sanitizes the names of sampled hyperparameters. `generate()` then evaluates the component effects used by the user input formula, in the context of the `newdata` values with the original component input specifications. It also detects it any `*_eval()` calls are used, allowing the effects to be evaluated for different specifications than the original component specifications (e.g. for an original spatial covariate `covareffect(eval_spatial(covar, where=geometry), ...)` one can use `covareffect_eval(newcovariatevalues)` to get the effect as a function of actual covariate values instead of a function of spatial location). The formula evaluation also has access to `*_latent` for each component, containing the raw latent state vector.
You can do any R evaluations in the expression, including indexing and function calls.

After `generate()` has transformed each sample by evaluating the "formula" expression, predict just does basic summary statistics on the result.

**Comment by tmeeha** [comment]
That is so cool! So now I am doing:

**make data with 58 states times 2 years = 116 rows and state_idx, year_idx, and log_hours columns**

test1 <- cbc2 %>% 
  select(state_idx, year_idx, state_hours, log_hours) %>% 
  distinct() %>% 
  group_by(state_idx) %>% 
  mutate(log_hours=log(mean(state_hours))) %>% 
  ungroup() %>% 
  select(-state_hours) %>% 
  filter(year_idx %in% c(1,10))

![image](https://github.com/inlabru-org/inlabru/assets/9539880/bf30704c-dee5-436a-807a-cb6f513df0d5)

**predict from data to get one difference between two big sums across states, one from each of two years**
test2 <- predict(newdata=test1,
    object=res1, 
    formula = ~ {
      sum(exp(Intercept + effort_eff + year_eff_eval(10) + 
            state_int + state_effort_dev + state_year_dev)) - 
      sum(exp(Intercept + effort_eff + year_eff_eval(1) + 
              state_int + state_effort_dev + state_year_dev))
    }, 
    n.samples=5000
)

I now get 1 row out with the mean difference and other summary stats. If I understand correctly, I needed to feed the predict function data with the years I wanted to evaluate, i.e., 1 and 10. Then in the formula for the predict function I took the two sums across states, each evaluated at a different year level (e.g., year_eff_eval(10)), and got the difference.  The difference (-33 million) is a little different from when I compute state by year predictions (see original question) and sum and difference them manually (-30 million) but maybe that is a sampling thing? Or an exponentiation before summing thing? Or an add variance components thing. Anyway, hopefully I am on the right track. Thanks for the advice, Finn!

**Reply by finnlindgren** [reply]
Yes, likely the difference between $E(\exp(x))$ and $\exp(E(x))$.  You can also see the Monte Carlo error in the `predict()` output; there's one column for the MC std.err. for the posterior mean, and one for the MC std.err. for the posterior standard deviation.

**Comment by tmeeha** [comment]
And thanks for the update on the SVC tutorial! There has been a lot of change in R spatial tools, so I appreciate you updating things. I'd like to make another tutorial some time on more basic use of inlabru for GLMMs that includes stuff about predicting and plotting from models.

**Reply by finnlindgren** [reply]
Yes, some things are available in the various vignettes, especially the lgcp vignettes, but we need more conceptual vignettes; we've started one on component definitions (needs some updating but is available), but more things are needed.
